src/drv_oculus_rift/rift-hmd-radio.c

```diff
@@ -0,0 +1,295 @@
+/*
+ * Oculus Rift CV1 Radio
+ * Copyright 2016 Philipp Zabel
+ * Copyright 2019 Jan Schmidt
+ * SPDX-License-Identifier:	BSL-1.0
+ */
+#include <stdint.h>
+#include <string.h>
+
+#include "rift-hmd-radio.h"
+#include "../ext_deps/nxjson.h"
+
+static int get_feature_report(hid_device *handle, rift_sensor_feature_cmd cmd, unsigned char* buf)
+{
+	memset(buf, 0, FEATURE_BUFFER_SIZE);
+	buf[0] = (unsigned char)cmd;
+	return hid_get_feature_report(handle, buf, FEATURE_BUFFER_SIZE);
+}
+
+static int send_feature_report(hid_device *handle, unsigned char* buf, int length)
+{
+	return hid_send_feature_report(handle, buf, length);
+}
+
+
+static bool rift_hmd_radio_send_cmd (hid_device *handle, uint8_t a, uint8_t b, uint8_t c)
+{
+	unsigned char buffer[FEATURE_BUFFER_SIZE];
+	int cmd_size = encode_radio_control_cmd(buffer, a, b, c);
+	int ret_size;
+
+	if (send_feature_report(handle, buffer, cmd_size) < 0)
+		return false;
+
+	do {
+		ret_size = get_feature_report(handle, RIFT_CMD_RADIO_CONTROL, buffer);
+		if (ret_size < 1) {
+			LOGE("HMD radio command 0x%02x/%02x/%02x failed - response too small", a, b, c);
+			return false;
+		}
+	} while (buffer[3] & 0x80);
```

Comment:
Should there be a timeout here?

Reply:
You tell me! This was lifted almost verbatim from ouvrt

In practice, the loop times out on its own if the controller doesn't respond - the high bit buffer[3] gets cleared and the error bit gets set.